### ISSUE TITLE
Unable to use multiple extensions

### ISSUE THREAD

**Issue by alicerunsonfedora** [migueldeicaza/SwiftGodot#65]
I am unable to tell if this is an issue with Godot itself or with the SwiftGodot library, but it seems that I can't write multiple extensions and have them load in the editor. It will pick one extension and load that, but not load the other.

**Steps to reproduce:**
1. Clone https://github.com/Indexing-Your-Heart/game/tree/00b21d34e8c01dafb4369825e157bc46952fd2e3 and run `just build-all-deps`.
2. Open the Godot editor by running `just edit-game`.
3. Create a new scene and open the new/add node dialogue.
4. Search for ProtractorDrawer and AnthroCharacterBody2D.
5. Note that while ProtractorDrawer appears, AnthroCharacterBody2D does not.
6. Close Godot and open Shounin/.godot/extension_list.cfg. Swap the two lines and save.
7. Re-open Godot and repeat step 4. Note that now ProtractorDrawer will disappear, but AnthroCharacterBody2D will appear.

**Expected behavior:**  
Both nodes appear in the node search list (i.e., both extensions are loaded).

**Comment by alicerunsonfedora** [comment]
**Update**: I think this is a library issue, since I tried throwing in the https://godotsteam.com/howto_gdextension/, and I'm able to access everything inside this extension such as `Steam.isSteamRunning()`.

**Comment by migueldeicaza** [comment]
thanks for the report I’ll investigate

**Comment by migueldeicaza** [comment]
Mhm, I have been struggling to get this built.

While the compilation ends with:

```
GenerateDSYMFile /Users/miguel/cvs/game/AnthroBase/xcbuild/Build/Products/Release-iphoneos/AnthroBase.framework.dSYM /Users/miguel/cvs/game/AnthroBase/xcbuild/Build/Products/Release-iphoneos/PackageFrameworks/AnthroBase.framework/AnthroBase (in target 'AnthroBase' from project 'AnthroBase')
    cd /Users/miguel/cvs/game/AnthroBase
    /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/dsymutil /Users/miguel/cvs/game/AnthroBase/xcbuild/Build/Products/Release-iphoneos/PackageFrameworks/AnthroBase.framework/AnthroBase -o /Users/miguel/cvs/game/AnthroBase/xcbuild/Build/Products/Release-iphoneos/AnthroBase.framework.dSYM

RegisterExecutionPolicyException /Users/miguel/cvs/game/AnthroBase/xcbuild/Build/Products/Release-iphoneos/PackageFrameworks/AnthroBase.framework (in target 'AnthroBase' from project 'AnthroBase')
    cd /Users/miguel/cvs/game/AnthroBase
    builtin-RegisterExecutionPolicyException /Users/miguel/cvs/game/AnthroBase/xcbuild/Build/Products/Release-iphoneos/PackageFrameworks/AnthroBase.framework

Touch /Users/miguel/cvs/game/AnthroBase/xcbuild/Build/Products/Release-iphoneos/PackageFrameworks/AnthroBase.framework (in target 'AnthroBase' from project 'AnthroBase')
    cd /Users/miguel/cvs/game/AnthroBase
    /usr/bin/touch -c /Users/miguel/cvs/game/AnthroBase/xcbuild/Build/Products/Release-iphoneos/PackageFrameworks/AnthroBase.framework

Resolve Package Graph
** BUILD SUCCEEDED **
```

The files listed there (like `/Users/miguel/cvs/game/AnthroBase/xcbuild/Build/Products/Release-iphoneos/PackageFrameworks/AnthroBase.framework`) do not exist.   Nor does the `/Users/miguel/cvs/game/AnthroBase/xcbuild` directory.  I am wondering what is going on here.

**Comment by alicerunsonfedora** [comment]
Hmm, I might've borked the build script. I'll try running it again and seeing what's up with this. I might need to set some new flags or something for the AnthroBase step.

**Comment by alicerunsonfedora** [comment]
Ah, wait a moment. @migueldeicaza Do the libAnthroBase.dylib and AnthroBase.framework extensions show up in `Shounin/bin`?

I just realized you might be looking for the build folder after it's finished building; by default, it cleans itself up and removes the `xcbuild` folder once the libraries are copied over.

**Comment by alicerunsonfedora** [comment]
I just did a clean build again, making sure I removed the older copies, and they do seem to be appearing in `Shounin/bin` as expected:
![image](https://github.com/migueldeicaza/SwiftGodot/assets/13445064/0aea4510-e37a-466c-be51-dcccd09e261f)

**Comment by migueldeicaza** [comment]
Thank you, there they are!   I was looking for libAnthroBase.dylib, but this one is already renamed to AnthroBase

**Comment by migueldeicaza** [comment]
Mhm, it looks like `AnthroBase` was intended for iOS (and I do have an ARM64 binary for that), but the `libAnthroBase.dylib` which is what the `.gdextension` file references it not there.

**Comment by alicerunsonfedora** [comment]
The framework files are for the iOS builds, and the dylibs are for the macOS versions (should be universal/fat binaries). Though I'm not certain why they wouldn't appear.

Maybe try force-building the Mac versions with:
- `just build-dep '-d -t mac -l ProtractorGodotInterop' Protractor`
- `just build-dep '-d -t mac' AnthroBase`

And seeing what comes out the other end? It's supposed to be building for both arm64 and x86_64 macOS, putting them into a single dylib.

**Comment by migueldeicaza** [comment]
Ok, that built, some observations:

* It did not work out of the box, I had to copy several files to new destinations, this is an area that I clearly would like to make it simpler to address, so things work out of the box.
* There are two separate copies of libSwiftGodot.dylib compiled, and I suspect that loading two into the same process is the source of the problems, but could not get there yet - will continue investgating tonight


I wonder if you can try using the .xcframework, which hopefully should just have a single copy of this library (I have just made the process easier, check the updated README)

**Comment by migueldeicaza** [comment]
Oops, reopening, I closed by accident.

**Comment by alicerunsonfedora** [comment]
> I wonder if you can try using the .xcframework, which hopefully should just have a single copy of this library (I have just made the process easier, check the updated README)

I'd love to use the xcframework going forward, but the current iterations still don't support M1 Macs (i.e., only contains macOS libraries for the x86_64 variant, no arm64 versions).

**Comment by migueldeicaza** [comment]
Thanks for catching this.

I have been building both binaries, but I must have messed something up, because only the Intel ones are shipping, debugging.

**Comment by alicerunsonfedora** [comment]
> Thanks for catching this.
> 
> I have been building both binaries, but I must have messed something up, because only the Intel ones are shipping, debugging.

I have something working on my end that I'm about to push into my fork. Might make a PR for it shortly.

**Comment by migueldeicaza** [comment]
oh, I was doing the iOS ones, not the Mac ones, ok that explains my confusion.

**Comment by alicerunsonfedora** [comment]
Alright, just opened the PR for the arm64 support: #70

**Comment by alicerunsonfedora** [comment]
Update: I've just tried this using a prebuilt xcframework, but it still appears to run into the exact same issue.

**Comment by migueldeicaza** [comment]
Ok, the issue was that every time we called the "initializeSwiftModule", this would overwrite the callback to register the previous object (at least with my custom setup I came up with). 

The code should now iterate over all the register ones.

That said, there is one area of concern and it is that there is a "token" that we are supposed to use with a number of Godot APIs which represent the GDExtension, and many of the registration APIs rely on this (ClassInfo) as well as the type instance creation and I suspect that not using this token correctly will lead to a crash (I imagine that they will use this to associate data with a library).

So the problem becomes: how to associate this token with the code.   

I suspect initializeSwiftModule can return a token that can be used for `ClassInfo` invocations - so this works, but I do not currently have a solution for associating a type at construction time with the correct library (this happens in Wrapped.swift), and I am not sure how would I pass this information down or retrieve this information at this point.

I need to think about that.
